tcpdump.c

```diff
@@ -0,0 +1,239 @@
+// snapshot len should then be MAX_PKT_LEN -1 
+#define MAX_PKT_LEN (65537)
+// TODO: fix, this is kinda random
+#define MAX_BPF_LEN (50)
+
+static bool running = true;
+
+void sigint_handler(int sig) {
+
+	running = false;
+    return;
+
+}
+
+static error_status_t set_bpf(int sock, int link_type, const char* bpf_str, int opt) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    struct sock_fprog bpf_filter = {0};
+    struct sock_filter* bpf_buffer = NULL;
+    
+    CHECK(sock != -1);
+    CHECK(bpf_str != NULL);
+
+    bpf_buffer = malloc(sizeof(struct sock_filter) * MAX_BPF_LEN);
+    CHECK(bpf_buffer != NULL);
+
+    bpf_filter.filter = bpf_buffer;
+
+    CHECK_FUNC(compile_bpf(MAX_PKT_LEN - 1, link_type, &bpf_filter, bpf_str, opt));
+
+    // for debug purpose
+    // the 2 stands for how i print the bpf 
+    dump_bpf(&bpf_filter, 2); 
+    CHECK(setsockopt(sock, SOL_SOCKET, SO_ATTACH_FILTER, &bpf_filter, sizeof(bpf_filter)) >= 0);
+
+cleanup:
+    free(bpf_buffer); // Best effort.
+    return ret_status;
+}
+
+static error_status_t iface_name_to_index(int sock, const char* if_name, int* if_index) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    struct ifreq ifr = {0};
+    size_t if_name_len = 0;
+
+    CHECK(if_name != NULL);
+    CHECK(if_index != NULL);
+    
+    if_name_len = strlen(if_name);
+
+    CHECK_STR(if_name_len < sizeof(ifr.ifr_name), "Interface name too long");
+    memcpy(ifr.ifr_name, if_name, if_name_len);
+    // ioctl for mapping if_name to if_index
+    CHECK(ioctl(sock, SIOCGIFINDEX, &ifr) != -1);
+
+    *if_index = ifr.ifr_ifindex;
+
+cleanup:
+    return ret_status;
+}
+
+static error_status_t iface_set_promisc(int sock, const char* if_name, bool set) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    int if_index = -1;
+    int opt_name = -1;
+    struct packet_mreq mreq = {0};
+    
+    CHECK(if_name != NULL);
+    CHECK_FUNC(iface_name_to_index(sock, if_name, &if_index));
+
+    mreq.mr_ifindex = if_index;
+    mreq.mr_type = PACKET_MR_PROMISC;
+
+	opt_name = set ? PACKET_ADD_MEMBERSHIP : PACKET_DROP_MEMBERSHIP;
+    CHECK(setsockopt(sock, SOL_PACKET, opt_name, &mreq, sizeof(mreq)) != -1);
+    // TODO: this is a side effect. what if i don't want to print?
+    printf("set interface: %s to promisc %d\n", if_name, set);
+
+cleanup:
+    return ret_status;
+}
+
+static error_status_t bind_raw_socket(int sock, const char* if_name) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    int if_index = -1;
+    struct sockaddr_ll sll = {0};
+    
+    CHECK(if_name != NULL);
+    CHECK_FUNC(iface_name_to_index(sock, if_name, &if_index));
+
+    // to bind a raw socket to a specific interface 
+    sll.sll_family = AF_PACKET;
+    sll.sll_protocol = htons(ETH_P_ALL);
+    sll.sll_ifindex = if_index;
+
+    CHECK_STR(bind(sock, (struct sockaddr*)&sll, sizeof(sll)) >= 0, "Failed binding to interface");
+
+cleanup:
+    // TODO: i don't close here even if it fails because i want main func to close it? 
```

Comment:
That's true. There's no reason for this function to close the socket, it didn't create it and doesn't own it.